### ISSUE TITLE
start command with :adjust does not overwrite open interval

### ISSUE THREAD

**Issue by xerus2000** [GothenburgBitFactory/timewarrior#326]
I already had a task running,but I realised it was the wrong one. It was started 10:50:05, so I typed

    ❯ timew start 10:50 mixxx
    The end of a date range must be after the start.

That error is already rather unhelpful, as it doesn't mention that the clash is due to the previous task.
But the more annoying problem is that even when specifying the adjust hint, I got that same error. I would expect to simply override the previous interval if that hint is supplied.

**Comment by lauft** [comment]
When you call `start` with an open interval in the database, Timewarrior attempts to close that interval at the given start time. In your case, you are trying to close an interval started at `10:50:05` at `10:50:00` which is an invalid interval and thus causes this error.

The adjust hint only tells Timewarrior to overwrite any intervals the new *valid* interval overlaps with. 

In your case you could have cancelled the wrong time tracking and start the correct one:
```
timew start WRONGTASK # 10:50:05
timew cancel
timew start RIGHTTASK 10:50
```

**Comment by xerus2000** [comment]
ah thanks, didn't know about the cancel command!

**Comment by sruffell** [comment]
As an aside, it doesn't *seem* unreasonable to me that the start command, when run with the :adjust hint, shouldn't be able to completely remove an interval that overlaps.

Would a patch that makes it behave this way be unwelcome?

**Comment by lauft** [comment]
The start command already adheres the `:adjust` hint when it comes to closed intervals:
```
$ timew track 19:00 - 20:00 FOO
...
$ src/timew start 18:00 BAR
You cannot overlap intervals. Correct the start/end time, or specify the :adjust hint.
$ src/timew start 18:00 BAR :adjust
...
$ src/timew summary 
Wk  Date       Day Tags    Start End    Time   Total
W21 2020-05-19 Tue BAR  18:00:00   - 2:37:19 2:37:19
                                                    
                                             2:37:19
```
As @xerus2000 pointed out, overwriting the open interval `BAR` like this fails:
```
$ src/timew start 17:00 BAZ :adjust
The end of a date range must be after the start.
```

To fix this I think the command `timew start <from>` should behave like this:
* Request all intervals that overlap with the range `[<from> - )` from the database
  This gives a chronological list of zero or many intervals, first being oldest, last being latest
* If the list is not empty:
    * If `<from>` is **within** the first interval:
        * If the first interval is open or the `:adjust` hint is given:
            * Truncate the first interval to `<from>`
            * Delete the remaining intervals of that list from the database
        * Else abort the command with an error message
    * Else (`<from>` is **before** the first interval):
        * If the `:adjust` hint is not given, abort the command with an error message
        * Else delete the intervals of that list from the database
* Finally, add the new interval to the database

The error message would be the usual message:
```
You cannot overlap intervals. Correct the start time, or specify the :adjust hint.
```

**Comment by sruffell** [comment]
Feel free to assign this to me if you would like me to submit a PR that implements this.

**Comment by lauft** [comment]
Implementation should be fairly easy. I would migrate the validation code between `CmdStart.cpp:52` and `CmdStart.cpp:91` to `autoAdjust` (`src/validate.cpp:98`) and implement a "standard interval overlap resolution" there, which handles the case of an open interval being added into an open interval. All other cases (which need the `:adjust` hint) should already be handled by the "overwrite resolution" (from `src/validate.cpp:116` and onward).

This way the start command merely creates the new open interval and adds it to the database. The rest happens during overlap resolution.

As this is a central location in the code, we must take care not to alter the behaviour of the other (range modifying) commands with the new introduced path for open intervals. But I think that is already covered.

As there is no way to modify existing intervals in the database into open intervals and the only place for an open interval in the database being the first one, the only way to trigger the new overlap resolution is to add a new open interval via `timew start`. That means moving an interval into the open interval (via `timew move`) or creating a closed interval within the open interval (via `timew track`) will still require the `:adjust` hint.

**Comment by sruffell** [comment]
This makes sense to me. I'll take a crack this weekend if nobody gets to it before then.

**Comment by sruffell** [comment]
actually, just looking at your comments again, I see that I need to migrate more into the autoAdjust function.  So, feel free to take a look, but I will most likely need to update the pull request.

**Comment by lauft** [comment]
@sruffell thanks again for your contribution, sorry my review took so long.

While reviewing I took the liberty to complete the migration you mentioned. I also moved the handling of adding an interval with the same tag set into the overlap resolution - `CmdStart` looks pretty clean now. :grin:

`CmdStart` is now also restricted to accept only a single datetime (or an open range) which means I dropped the redirect to `CmdTrack` when called with a (closed) range. Instead a message is displayed suggesting `CmdTrack`(just like `CmdStop` suggesting `CmdModify` when called with an id).
